add code to measure memory usage

pmnoxx opened the pull request:
draft only

bowenwang1996 commented on chain/chain/src/test_utils.rs:
```diff
@@ -44,18 +45,29 @@ use crate::types::{ApplyTransactionResult, BlockHeaderInfo, ChainGenesis};
 
 #[derive(
-    BorshSerialize, BorshDeserialize, Serialize, Hash, PartialEq, Eq, Ord, PartialOrd, Clone, Debug,
+    BorshSerialize,
```
we don't need to add this to test-only structs

pmnoxx replied:
We need it for KeyValueRuntime in order for it to implement RuntimeAdater trait.

If needed I can change the code, so we don't do it for structures used by KeyValueRuntime,

bowenwang1996 commented on Cargo.toml:
```diff
@@ -75,13 +77,16 @@ neard = { path = "./neard"}
-lto = true        # Enable full link-time optimization.
-codegen-units = 1 # Use only 1 codegen-unit to enable full optimizations.
+lto = false        # Enable full link-time optimization.
+codegen-units = 256 # Use only 1 codegen-unit to enable full optimizations.
```
why do we change these two configs here?

pmnoxx replied:
I disabled those settings to decrease build time from 5m to 17s on my local machine. Mozilla was dealing with similar issue for Firefox, they ended up disabling lto for dev machines, but they enabled it only for release builds:
https://bugzilla.mozilla.org/show_bug.cgi?id=1386371
here is their PR - https://hg.mozilla.org/integration/autoland/rev/d75f07915a48
I think we should do something similar.

bowenwang1996 commented on Cargo.toml:
```diff
@@ -75,13 +77,16 @@ neard = { path = "./neard"}
 
 [profile.bench]
-lto = true
+lto = false
```
also here

pmnoxx replied:
fixed

bowenwang1996 commented on chain/chain/src/chain.rs:
```diff
@@ -173,6 +176,8 @@ impl OrphanBlockPool {
 
 /// Facade to the blockchain block processing and storage.
 /// Provides current view on the state according to the chain state.
+/// #[
```
?

pmnoxx replied:
fixed

bowenwang1996 reviewed (commented):
Looks good. I think after we validated that it works, some of the next steps I see are:
1) put memory tracker related code under a compile-time flag
2) extract deepsize out (probably to its own repo)

pmnoxx commented:
> Looks good. I think after we validated that it works, some of the next steps I see are:
> 
> 1. put memory tracker related code under a compile-time flag
I also need to change how the data is presented to be more readable.
> 2. extract deepsize out (probably to its own repo)
Is there a way to extract a package into it's own repo, without making crate public on crates.io.
Or do you think it's fine to release a new package for crates.io?


I'm planning to add more things to this PR after I get initial results about the memory usage. It really depends on where the memory leak is located.

bowenwang1996 commented:
@pmnoxx why did you close this PR?

pmnoxx commented:
I merged https://github.com/near/nearcore/pull/3747
In that change we added memory profiling, as well as printing memory usage per thread.

We already moved  memory tracker related code to it's own repo.

I'll keep this issue open. I forgot I still need to create a change for deepsize package.

bowenwang1996 commented:
@pmnoxx I see. Feel free to close this PR then
